src/Microsoft.ML.Data/Transforms/Hashing.cs

```diff
@@ -490,34 +605,79 @@ public uint HashCore(uint seed, uint mask, in double value)
             }
         }
 
+        private readonly struct HashDoubleV2 : IHasher<double>
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public uint HashCore(uint seed, uint mask, in double value)
+            {
+                if (double.IsNaN(value))
+                    return 0;
+
+                ulong v = FloatUtils.GetBits(value == 0 ? 0 : value);
+                var hash = Hashing.MurmurRound(seed, Utils.GetLo(v));
+                var hi = Utils.GetHi(v);
+                if (hi != 0)
+                    hash = Hashing.MurmurRound(hash, hi);
+                return (Hashing.MixHashV2(hash, 4) & mask);
```

Comment:
>4 [](start = 48, length = 1)

What does this number signify? Is it the number of bytes of the input?
I see that it is always called with the value 4, should we consider making the public signature of this method just `MixHashV2(hash)`, and then have it call a private method that takes two parameters?

Reply:
MixHashV2 is called with a different length when called from MurmurHashV2 for strings. So I think it should remain the way it is. 

---
In reply to: [415252298](https://github.com/dotnet/machinelearning/pull/5013#discussion_r415252298) [](ancestors = 415252298)

Reply:
The difference between `MixHash` and `MixHashV2` is only this line:
```
hash ^= (uint)len;
```
and it is called with the constant value `4` everywhere except for the `MurmurHashV2` for strings. I understand the purpose of somehow mixing in the length of the string when hashing string, but what value does it add to always mix in `4`? Is this how ONNX runtime hashes uints? 

---
In reply to: [416079308](https://github.com/dotnet/machinelearning/pull/5013#discussion_r416079308) [](ancestors = 416079308,415252298)

Reply:
Yes, this is how ORT hashes uints. 
Please [see here](https://github.com/microsoft/onnxruntime/blob/78fde2c4cbeaaaf46fa97e1ffaea0fa01cbbcb33/onnxruntime/contrib_ops/cpu/murmur_hash3.cc#L155).

This is part of the reason why we had to redo our hashing so that we can be aligned with ORT.

Reply:
Thanks for providing the link to the ORT implementation. `MurmurHash3_x86_32` is called in two places: when the input is a string (in which case, the length of the string is passed as `len`), and when the input is a number (in which case the number of bytes is passed as `len`). In the numeric case, ORT has a check that verifies that the number has 4 bytes, but the value passed to `MurmurHash3_x83_32` is still the element size, and not the constant `4`.

If we want to truly be consistent with ORT hashing (and actually, with the `MurmurHash3_x86_32` algorithm), I think we should also use `sizeof` instead of the constant `4` when calling `MixHash`. We should definitely do that for `long`, `ulong` and `double` (which cannot be exported to ONNX anyway). However, if we do that for types with less than 4 bytes, the values computed by the exported model as it is written in this PR will not be consistent with the ML.NET values (because of the additional "Cast" operator added to convert the original input type to `uint`). This means that we would only be able to export to ONNX models with inputs that are either ints, uints or strings. To me that is a reasonable price to pay, since I don't think it is common to have pipelines where people are trying to hash bytes/shorts. What do you think @harishsk, @Lynx1820 ?



---
In reply to: [416808389](https://github.com/dotnet/machinelearning/pull/5013#discussion_r416808389) [](ancestors = 416808389)

Reply:
The current implementation is consistent with ORT.  @Lynx1820 has tests in OnnxConversionTest.cs for all data types (except long, ulong, double and float). 

You are correct that when `MurmurHash3_x86_32` is called the `len` parameter is the size in bytes. This in turn gets translated to a `nblocks` value of `0`. This causes the code in lines 122 to 131 to be executed only once. But on line 120, the data `data` gets cast to a pointer of blocks of type `uint32_t` and it hashes each block as a `uint32_t` value.

The MurmurHash3 algorithm is defined in terms of hashing everything as `uint32_t` blocks. So to be consistent, it is the right thing to do always hash all types of size <= 4 bytes as `uint32_t` types.  We are already doing that and the tests are passing with consistent hashing values between ML.NET and ORT.

I do agree with you about the constant 4 and that it should be expressed in more generic terms. I think it should be changed to `sizeof(uint)` to make it explicit that everything is being hashed as a `uint`. 

@Lynx1820 can you please make that change?

@yaeldekel would that fix be sufficient to resolve this comment?




---
In reply to: [417988309](https://github.com/dotnet/machinelearning/pull/5013#discussion_r417988309) [](ancestors = 417988309,416808389)

Reply:
I think for now this is sufficient. There is still a question about the 8 byte types in case ORT decides to support that in the future - but since we don't know how they would choose to do it, the approach we take is fine.

---
In reply to: [418289134](https://github.com/dotnet/machinelearning/pull/5013#discussion_r418289134) [](ancestors = 418289134,417988309,416808389)